tests/test-images/help.1
.TH "ATOMIC" "1" " Atomic Man Pages" "Brent Baude" "December 2015"  ""


.SH DESCRIPTION
.PP
Simple \ftest\fP case

.TS
allbox;
l l
l l .
Simple	Table
Simple	Cell
.TE

.SH USAGE
.PP
test

.PP
.RS

.SH HISTORY
.PP
December 2015, Originally written by Brent Baude (bbaude at redhat dot com)


CHANGELOG.md
## 1.17 (2017-04-20)
    Allow anonymous push
    Always use fq names for dest image
    dbus: Lots of fixes
    Atomic/install.py: Record installs for later use
    Signing
    Fix registries.d/*.yaml config lookup
    Fix “fqdn” computation in RegistryInspect
    Honor proxy usage
    Makefile: add new rule test-suite
    Only compute util.Decompose(expanded_image_name) once
    Simplify registries.d/*.yaml config lookup
    Update (atomic sign) for the updated sigstore layout
    atomic.d/openscap: Improve openscap scans
    backends: add tag_image method
    bash: add autocompletion for images tag
    docs: document images tag
    image: add new subcommand tag
    install: add --system-package
    Improve Atomic Install
        Lots of improvements in system containers
        syscontainer: support rpm installs
	syscontainers: Add run once logic to install
    syscontainers: fix containers list --json
    syscontainers: use --systemd-cgroup with runC
    systemcontainers: fill default values when using oneshot
    util: Don't try to create directories at module import time

## 1.16 (2017-02-22)
    Fix uninstall bug: BZ 1425495
    add --lvsize option to atomic storage modify
    Account for API Changes in docker-py-2
    Suppress STDOUT of lower level commands e.g. vgreduce, wipefs.
    Fix Stop Regression: BZ #1422448
    Enable dbus install, stop.
    Allow for async scans via dbus
    syscontainers: allow delete by image id
    syscontainers: support @sha256: format for image listing
    Fix FQ Name for SystemContainers

## 1.15 (2017-01-17)
    update default trust policy file
    Validate reg input to trust add cmd
    Atomic/objects/image.py: Fix verify for v1
    Atomic/verify.py: Fix dbus implementation of image verify
    main: Don't catch all AttributeErrors
    storage: Process arguments in set_args, not __init__
    images, containers: do lowercase comparison for filter values
    images: apply filters before any output
    containers: --json exports image_id
    Refactor atomic stop
    Add keywords to completions

## 1.14 (2017-01-02)
    update: refactor into non-base verbs
    Atomic storage reset does not work on docker-latest
    Fixes for documentation
    syscontainers: unlink temporary file if substitution fails
    syscontainers: simplify substitution of variables
    Add --all to images delete
    tests: replace sed with a python script
    syscontainers: prune the ostree repo with images prune
    syscontainers: allow delete multiple images by ID
    syscontainers: use the image id from the raw manifest
    syscontainers: use system checkout as import tmp directory
    syscontainers: fix tarfile import with no RepoTags
    syscontainers: generate an UUID at installation time
    syscontainers: update honor --force
    Refactor containers verb
    Atomic/diff.py: Fix options bug
    Unify and refactor atomic verify
    fix get auth from docker.io
    redhat-ci: make testsuites required
    push: prompt user/pass lowercase
    Atomic/diff.py: Use go-mtree for file comparisons
    run: add --detach and only add -t if in a TTY
    dbus: fix Install() and Run() signatures
    pull: support dockertar for docker backend
    Atomic/top.py: Fix options handling in top
    generate: default storage for mounts
    Add fedora25_cloud target for vagrant
    Atomic Info Unittests
    Refactor images
    The HELP label by default should be "help"
    atomic_dbus: keep the name until the process exits
    Add substitutions for Opt variables
    Minor fix to delete
    Atomic/mount.py: Re-Add _clean_temp_container_by_path (BZ 1397839)
    test_util.py: adapt for newer sepolicies
    syscontainers: add rollback
    backends: has image|container return objects
    syscontainers: fix installation
    backends: add skeleton for ostree backend
    syscontainers: allow to specify what image to pull
    syscontainers: get_containers accept what containers to inspect
    Add refactoring structure
    Atomic/mount.py: shutil.rmtree input must be dir
    Use centos for all test images
    syscontainers: output better json errors
    atomic diff: Add ability to compare metadata
    syscontainers: environment variable detection
    Add  SYSTEMD_IGNORE_CHROOT=1 to environment of SPCs
    The HELP label by default should be "help"
    make vagrant-check: Run tests with vagrant

## 1.13 (2016-12-13)
Refactor verbs:
	containers
	update
	verify
	images
backends: has image|container return objects
Add  SYSTEMD_IGNORE_CHROOT=1 to environment of SPCs
Atomic diff: Use go-mtree for file comparisons
syscontainers:
	Lots of Bug Fixes
	simplify substitution of variables
	add rollback
Signing:
	push - use credentials in skopeo copy
	pull: support dockertar for docker backend
	fix get auth from docker.io
	prompt user/pass lowercase

Dbus Bindings:

## 1.13 (2016-10-24)
Add --storage option to image-related commands
syscontainers: Fix docker: and dockertar: installs
Atomic/images.py: Enable filter for dangling.
Add atomic trust verb
Add support for image signing
Add support for overlay2 driver
Allow pull from registry not in docker conf
Add dbus support for atomic stop
Add dbus support for atomic install/uninstall
Add dbus support for atomic run
Add dbus support for atomic pull
Add dbus support for atomic top.
Remove primary commands and move to images subcommand
    atomic help
    atomic info
    atomic verify
    atomic version
Introduce registry inspect methods

## 1.12 (2016-9-7)
Fixes for syscontainers
Add atomic images generate to generate mtree meta data
Fix up atomic with overlay backend
Add atomic sign to allow simple signing of images
Add atomic pull support for signatures

## 1.11 (2016-8-9)
Add support for system containers
Add support for managing storage
Add support for atomic ps
Improve dbus interfaces
Improvements to atomic scan

## 1.10 (2016-5-25)
Improve Error Handling
- Unify error messages for no docker daemon (BZ #1300187)

Add atomic storage command

Atomic/help.py
from . import Atomic
import argparse
import tempfile
from pydoc import pager
import os
from . import mount
import sys
from . import util

def cli(subparser, hidden=False):
    #atomic help
    if hidden:
        helpp = subparser.add_parser("help", argument_default=argparse.SUPPRESS)

    else:
        helpp = subparser.add_parser("help",
                                     help=_("display help associated with the image"),
                                     epilog="atomic help 'image'")

    helpp.set_defaults(_class=AtomicHelp, func='help_tty')
    helpp.add_argument("image", help=_("Image ID or name"))


GROFF_BINARY = "/usr/bin/groff"


class AtomicHelp(Atomic):
    ATOMIC_DIR="/run/atomic"

    def __init__(self):
        super(AtomicHelp, self).__init__()
        if not os.path.exists(self.ATOMIC_DIR):
            os.makedirs(self.ATOMIC_DIR)

        # possible filenames for help file inside container
        self.help_file_candidates = (
            # help filename inside image, command to process the help file before displaying
            ("help.1", (GROFF_BINARY, '-t', '-man', '-Tascii')),
            ("README.md", None)
        )

        self.docker_object = None
        self.is_container = True
        self.use_pager = True
        self.image = None
        self.inspect = None
        self.enc = sys.getdefaultencoding()

    def help_tty(self):
        result = self.help()
        if not sys.stdout.isatty():
            util.write_out("\n{}\n".format(result))
        else:
            # Call the pager
            os.environ['PAGER'] = '/usr/bin/less -R'
            pager(result)

    def help(self):
        """
        Displays help text for a container.
        :return: None
        """
        self.docker_object = self.args.image
        docker_id = self.get_input_id(self.docker_object)
        self.inspect = self._inspect_container(docker_id)
        if self.inspect is None:  # docker_id is an image
            self.inspect = self._inspect_image(docker_id)
            self.is_container = False
        else:
            # The docker object is a container, need to set
            # its image
            self.image = self.inspect['Image']

        try:
            return self.man_help(docker_id)
        except ValueError:
            # Check if "help" label is provided
            # The label contains command which will be executed
            help_cmd = self._get_args('HELP')
            if help_cmd:
                return self.alt_help(help_cmd)
            else:
                raise ValueError("There is no help for {}.".format(self.docker_object))

    def man_help(self, docker_id):
        """
        Display the help for a container or image using the default
        method of displaying a man formatted page
        :param docker_id: docker object to get help for
        :return: None
        """
        mount_location = tempfile.mkdtemp(prefix=self.ATOMIC_DIR)
        try:
            dm = mount.DockerMount(mount_location)
            with mount.MountContextManager(dm, docker_id):
                # defined b/c of pylint, this is not needed due to for & else construct
                candidate_file = None
                candidate_preprocessor = None
                for candidate_file, candidate_preprocessor in self.help_file_candidates:
                    # overlay
                    help_path = os.path.join(dm.mountpoint, candidate_file)
                    if not os.path.exists(help_path):
                        # devicemapper
                        help_path = os.path.join(dm.mountpoint, 'rootfs', candidate_file)
                    if os.path.exists(help_path):
                        break
                else:
                    # not found
                    raise ValueError(
                        "Unable to find help file for {}.\nTried these files {}.".format(
                            self.docker_object, [x[0] for x in self.help_file_candidates])
                    )

                with open(help_path, "r") as help_file:
                    if candidate_preprocessor:
                        if not os.path.exists(candidate_preprocessor[0]):
                            raise IOError(
                                "Cannot display help file {} for {}: {} unavailable".format(
                                    candidate_file, self.docker_object, candidate_preprocessor)
                            )

                        return util.check_output(candidate_preprocessor, stdin=help_file)
                    return help_file.read()
        finally:
            os.rmdir(mount_location)

    def alt_help(self, help_cmd):
        """
        Returns help when the HELP LABEL override is being used.
        :return: None
        """
        cmd = self.gen_cmd(help_cmd)
        cmd = self.sub_env_strings(cmd)
        self.display(cmd)
        return util.check_output(cmd, env=self.cmd_env()).decode(self.enc)
